clang/lib/AST/ASTContext.cpp

```diff
@@ -444,7 +444,7 @@ const RawComment *ASTContext::getRawCommentForAnyRedecl(
     return CommentlessRedeclChains.lookup(CanonicalD);
   }();
 
-  for (const auto Redecl : D->redecls()) {
+  for (const auto Redecl : CanonicalD->redecls()) {
```

Comment:
I reread your analysis and I think the only safe usage of `getRawCommentForAnyRedecl()` previously was that, in the second call to the function, only those `D`s that live in between the `CanonicalDecl` and `LastCheckedRedecl` are expected to be used as the starting point of the traversal. Otherwise, the presence of `LastCheckedRedecl` would result in skipping over all declarations, including those whose associated comments were unparsed yet.

So I wonder if we could make use of LastCheckedRedecl opt-in. To be clear, we first check if `D` is in the path `Canonical` -> `LastCheckedRedecl`. If so, we use the mechanism that skips past every visited declaration; otherwise, we ignore it.

```cpp
// Any redeclarations of D that we haven't checked for comments yet?
const Decl *LastCheckedRedecl = CommentlessRedeclChains.lookup(CanonicalD);
bool CanUseCommentlessCache = false;
if (LastCheckedRedecl) {
  for (auto *Redecl : CanonicalD->redecls()) {
    if (Redecl == D) {
      CanUseCommentlessCache = true;
      break;
    }
    if (Redecl == LastCheckedRedecl)
      break;
   }
}
```

```cpp
// Skip all redeclarations that have been checked previously.
if (CanUseCommentlessCache && LastCheckedRedecl)
  ...
```

The solution could be improved to pick up the `LastCheckedRedecl` again after visiting the `CanonicalDecl`, though.

Reply:
Thank you for the suggestion! It indeed feels safer to tweak `getRawCommentsForAnyRedecl()` like this than to muck around with the modeling of template specializations :laughing:

Reply:
Yeah, but I still feel the modeling of template specializations should be improved regarding `getSourceRange()` relying on `TemplateParameterListsInfo`, which is really peculiar. But that's another murky topic unrelated to this patch, anyway.

Reply:
> regarding `getSourceRange()` relying on `TemplateParameterListsInfo`, which is really peculiar

FWIW, that part actually makes sense to me. For a function/method declaration that looks like this:

```
template <...>
ReturnType FunctionName(Parameters...);
```

the answer to the question "where does the source range begin?" is indeed "if there is a template parameter list, where does it begin?"

Reply:
(The part that makes less sense to me is, why does the explicit specialization have a redeclaration at all in a case like

```c++
/// Aaa.
template<typename T, typename U>
void foo(T aaa, U bbb);

/// Bbb.
template<>
void foo(int aaa, int bbb);
```

There is only one declaration in the code.)

Reply:
Yup, in fact, I think we're not too confident about how many FunctionDecls would be generated during the template parsing e.g. they can be the byproducts of a TreeTransform/the process of the template argument deduction, such that the contract of `TemplateParameterListsInfo` would be less reliable.